### ISSUE TITLE
Modify service name base on tag/metadata

### ISSUE THREAD

**Issue by jzhao20230918** [hashicorp/consul#19751]
hello, during the register phase, is it possible that the service name be modified based on the tag/metadata？
I'm using Nomad to deploy service in canary mode, but the canary and the live instance would have the same service name in Consul and make it different in routing configuration. Any hint or help would be appreciated.

**Comment by huikang** [comment]
Consul allows you to split traffic under the same service name by `service-splitter`. See example https://developer.hashicorp.com/consul/docs/connect/config-entries/service-splitter#two-subsets-of-same-service

**Comment by next-jesusmanuelnavarro** [comment]
> Consul allows you to split traffic under the same service name by `service-splitter`. See example https://developer.hashicorp.com/consul/docs/connect/config-entries/service-splitter#two-subsets-of-same-service

Yes, you can split traffic but it seems you can't use that for traffic routing at the api gateway (see #22575) which means you basically negate the ability to use canaries for services directly exposed on it.

One workaround would be for the api gateway to support tag/meta-based routing (which is the scope of #22575); the other, is this issue's scope: having the ability to register canaries with a different service name for as long as they are active and transparently falling back to the "standard" service name once promoted (i.e.: without risking a service failure right after promotion).

Note both approaches are not incompatible and this one is somehow a more general one since it can be easily used by any reverse-proxy you may be using.
